static/app/api.tsx

```diff
@@ -668,31 +629,81 @@ export class Client {
     // or handle with a user friendly error message
     const preservedError = new Error('API Request Error');
 
-    return new Promise((resolve, reject) =>
+    return new Promise((resolve, reject) => {
       this.request(path, {
         ...options,
         preservedError,
-        success: (data, textStatus, resp) => {
+        error: (resp: ResponseMeta) => {
+          // Although `this.request` logs all error responses, this error object can
+          // potentially be logged by Sentry's unhandled rejection handler
+          reject(new RequestError(options.method, path, preservedError, resp));
+        },
+        success: (data: any, textStatus?: string, resp?: ResponseMeta) => {
           if (includeAllArgs) {
             resolve([data, textStatus, resp] as any);
           } else {
             resolve(data);
           }
         },
-        error: (resp: ResponseMeta) => {
-          const errorObjectToUse = new RequestError(
-            options.method,
-            path,
-            preservedError,
-            resp
-          );
+      });
+    });
+  }
 
-          // Although `this.request` logs all error responses, this error object can
-          // potentially be logged by Sentry's unhandled rejection handler
-          reject(errorObjectToUse);
+  handleRequestError(
+    options: {
+      id: string;
+      path: string;
+      requestOptions: Readonly<RequestOptions>;
+    },
+    response: ResponseMeta,
+    textStatus: string,
+    errorThrown: string
+  ) {
+    const code = response?.responseJSON?.detail?.code;
+    const isSudoRequired = code === SUDO_REQUIRED || code === SUPERUSER_REQUIRED;
+
+    let didSuccessfullyRetry = false;
+
+    if (isSudoRequired) {
+      openSudo({
+        isSuperuser: code === SUPERUSER_REQUIRED,
+        sudo: code === SUDO_REQUIRED,
+        retryRequest: async () => {
+          try {
+            const data = await this.requestPromise(options.path, options.requestOptions);
+            if (options.requestOptions.success) {
+              options.requestOptions.success(data, textStatus, response);
+            }
```

Comment:
nit: this could be done with optional-chaining on the function call, as it’s also done in the `error` case:

```suggestion
            options.requestOptions.success?.(data, textStatus, response);
```

Reply:
Changed it back. I originally modified it because it made it easier to wrap my head around the logic + I'm not really a fan of tiny character inside some long call sites like this that affect the execution (maybe I'm just getting old)